### ISSUE TITLE
Pretty print 5.2-compatible way

### ISSUE THREAD

**Issue by mnelson4** [WP-API/WP-API#1035]
I see #1027 and #835 were on this same topic. I have another pull request for the same feature, but different implementation that IS compatible with PHP 5.2.

**Comment by danielbachhuber** [comment]
I think this makes most sense in a supplemental plugin for now.

**Comment by mnelson4** [comment]
yeah ok although the wpcom API provides this.

If this is to be done by a supplemental plugin we should add a filter for it, or is there one already?

**Comment by nerrad** [comment]
I think it'd be valuable for debugging to have this included.  Especially when this eliminates the sole reason given in #835 for not doing something similar (PHP5.2).

**Comment by mnelson4** [comment]
Also thinking about it: an API client developer might not have access to the server to install another plugin, in which case they're going to need to go to some 3rd party service like @rmccue pointed out on #835. I think given the frequency this is requested merits NOT putting it in another plugin

**Comment by danielbachhuber** [comment]
> I think given the frequency this is requested merits NOT putting it in another plugin

But, it can easily be replicated with a browser extension, no?

**Comment by mnelson4** [comment]
I think the time required to merge this in is much less than the time it will take nearly ALL api clients to figure out how to pretty-up the json in their client (which of course won't always be a browser, or command line; but will often be javascript code, server-side code, etc).
for the event espresso 3 api, we included it because our iOS developer requested it, and it only took a few minutes for me to grab the snippet and implement it, rather than needing to explain all the different ways to pretty-up the JSON in your client to ALL the different developers.

**Comment by danielbachhuber** [comment]
This isn't a priority for us right now.

**Comment by mnelson4** [comment]
Ok so maybe I'll bring it up again in 2.0, at least adding a filter so a plugin can add this

**Comment by danielbachhuber** [comment]
> at least adding a filter so a plugin can add this

You can use the `json_pre_serve_request` filter: https://github.com/WP-API/WP-API/blob/develop/lib/infrastructure/class-wp-json-server.php#L281

**Comment by mnelson4** [comment]
Ya ok that works, and I have a plugin located here: https://github.com/eventespresso/wp-api-pretty-json/blob/master/wp-api-pretty-json.php
However, it seems like that short-circuits a lot more code than what's necessary in this case: my filter would need to basically duplicate the code between lines 283 and 307 (which is difficult because many of those methods are protected), but just pass the final result through the newly added function `prettyJson` method in the pull request. 
Basically, I'd want another filter before line 300, the check for `if ( isset( $_GET['_jsonp'] ) ) {...`, which would essentially act as another "ifelse" in that if-else statement. But no worries, I get that you guys are busy and aren't looking for enhancements

**Comment by rmccue** [comment]
> I think the time required to merge this in is much less than the time it will take nearly ALL api clients to figure out how to pretty-up the json in their client

I'd be happy to accept a great PR that adds this in, in a well-coded and well-formatted way that works across all versions of PHP that we support. It needs to use the built-in PHP version if available (for performance), or otherwise handle it itself.

Ideally, this actually needs to happen in `wp_json_encode`, WP's built-in compatibility shim. I pinged @pento (the steward of said function) about this, and he confirmed that he'd gladly accept a patch to do that. With support in `wp_json_encode`, adding it to the API is trivial. (I'm also happy to accept it into the API separately, and we can handle taking it upstream to WP core.)

So far, there's not been a PR that ticks all the boxes here. As @danielbachhuber noted, it's not a priority for us (given the abundance of external tools available), but that just means we can't dedicate a bunch of our own time to this. We're happy to review and accept PRs made against `develop` and integrate them if they're good, but it's not a core focus for us.
